src/Surfnet/StepupSelfService/SelfServiceBundle/Service/RemoteVetting/SamlCalloutHelper.php

```diff
@@ -0,0 +1,124 @@
+<?php
+/**
+ * Copyright 2010 SURFnet B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Surfnet\StepupSelfService\SelfServiceBundle\Service\RemoteVetting;
+
+use Psr\Log\LoggerInterface;
+use SAML2\Constants;
+use SAML2\XML\saml\SubjectConfirmation;
+use Surfnet\SamlBundle\Http\PostBinding;
+use Surfnet\SamlBundle\SAML2\AuthnRequestFactory;
+use Surfnet\StepupSelfService\SelfServiceBundle\Service\RemoteVetting\ValueObject\ProcessId;
+use Surfnet\StepupSelfService\SelfServiceBundle\Service\RemoteVetting\Dto\RemoteVettingTokenDto;
+use Surfnet\StepupSelfService\SelfServiceBundle\Service\RemoteVettingService;
+use Symfony\Component\HttpFoundation\Request;
+
+class SamlCalloutHelper
+{
+    /**
+     * @var IdentityProviderFactory
+     */
+    private $identityProviderFactory;
+    /**
+     * @var PostBinding
+     */
+    private $postBinding;
+    /**
+     * @var RemoteVettingService
+     */
+    private $remoteVettingService;
+    /**
+     * @var LoggerInterface
+     */
+    private $logger;
+    /**
+     * @var ServiceProviderFactory
+     */
+    private $serviceProviderFactory;
+
+    public function __construct(
+        IdentityProviderFactory $identityProviderFactory,
+        ServiceProviderFactory $serviceProviderFactory,
+        PostBinding $postBinding,
+        RemoteVettingService $remoteVettingService,
+        LoggerInterface $logger
+    ) {
+        $this->identityProviderFactory = $identityProviderFactory;
+        $this->serviceProviderFactory = $serviceProviderFactory;
```

Comment:
I've considered that but then a CompilerPass or something like that was needed for the initialization of the PrivateKey. So a lot more coupling would then be introduced. And I considered that ultimately didn't outweigh the readability and simplicity of the current solution.

Reply:
I think you do not understand what I'm impying here.

Proposed `__construct`
```
    public function __construct(
        IdentityProviderFactory $identityProviderFactory,
        ServiceProviderFactory $serviceProviderFactory,
        PostBinding $postBinding,
        RemoteVettingService $remoteVettingService,
        LoggerInterface $logger
    ) {
        $this->identityProviderFactory = $identityProviderFactory;
// Directly create the SP instead of exposing the SP factory        
$this->sp = $this->serviceProviderFactory->create();
        $this->postBinding = $postBinding;
        $this->remoteVettingService = $remoteVettingService;
        $this->logger = $logger;
    }
```

I don't see why you would need additional compiler passes for this? But maybe I'm wrong!

Reply:
I thought you meant with 'during construction time' the construction of the class when bootstrapping Symfony and not in the constructor of the object itself.

Reply:
Thanks for the update!